GCC 8.5 produces large force differences (at least in ppc64le)

I have been trying to build [conda-forge packages for PPC64LE](https://github.com/conda-forge/openmm-feedstock/pull/23) to see if it was possible after the last changes in `master` and feeling optimistic after seeing the Travis CI in this repo.

Anaconda provides the compilers for ppc64le too, but this is the GCC 8.x branch.  Apparently, this version (as opposed with 7.x) does not work well on, at least, `ppc64le`.

Check [logs for attempt here](https://travis-ci.com/github/jaimergp/openmm/builds/169335205). It contains builds for GCC 7.x (everything ok) and GCC 8.x (fails) under PPC64LE. Namely:

* `python -m simtk.testInstallation` fails with large force differences

```
OpenMM Version: 7.5
Git Revision: ef66223c78c6a416cb96ab802c71eca718051747
There are 2 Platforms available:
1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
Median difference in forces between platforms:
Reference vs. CPU: 0.194358  *** LARGE DIFFERENCE **
```

* pytest: 

```
__________________________ TestForceField.test_Forces __________________________
self = <TestForceField.TestForceField testMethod=test_Forces>
    def test_Forces(self):
        """Compute forces and compare them to ones generated with a previous version of OpenMM to ensure they haven't changed."""
    
        pdb = PDBFile('systems/lysozyme-implicit.pdb')
        system = self.forcefield2.createSystem(pdb.topology)
        integrator = VerletIntegrator(0.001)
        context = Context(system, integrator)
        context.setPositions(pdb.positions)
        state1 = context.getState(getForces=True)
        with open('systems/lysozyme-implicit-forces.xml') as input:
            state2 = XmlSerializer.deserialize(input.read())
        numDifferences = 0
        for f1, f2, in zip(state1.getForces().value_in_unit(kilojoules_per_mole/nanometer), state2.getForces().value_in_unit(kilojoules_per_mole/nanometer)):
            diff = norm(f1-f2)
            if diff > 0.1 and diff/norm(f1) > 1e-3:
                numDifferences += 1
>       self.assertTrue(numDifferences < system.getNumParticles()/20) # Tolerate occasional differences from numerical error
E       AssertionError: False is not true
TestForceField.py:298: AssertionError
=============================== warnings summary ===============================
/home/travis/virtualenv/python3.8.0/lib/python3.8/distutils/__init__.py:4
  /home/travis/virtualenv/python3.8.0/lib/python3.8/distutils/__init__.py:4: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp
TestAmberPrmtopFile.py::TestAmberPrmtopFile::testNucleicGBParametes
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolvent
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolventForces
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolventParameters
  /home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/simtk/openmm/app/internal/amber_file_parser.py:1018: UserWarning: Non-optimal GB parameters detected for GB model GBn
    warnings.warn(
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolvent
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolventForces
TestAmberPrmtopFile.py::TestAmberPrmtopFile::test_ImplicitSolventParameters
  /home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/simtk/openmm/app/internal/amber_file_parser.py:1018: UserWarning: Non-optimal GB parameters detected for GB model HCT
    warnings.warn(
TestForceField.py::TestForceField::test_ImpropersOrdering_smirnoff
  /home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/simtk/openmm/app/internal/pdbstructure.py:537: UserWarning: WARNING: duplicate atom (HETATM    2          A   1       0.513  -1.098  -0.004  1.00  0.00           O  , HETATM    1          A   1       0.095   0.011   0.000  1.00  0.00           C  )
    warnings.warn("WARNING: duplicate atom (%s, %s)" % (atom, old_atom._pdb_string(old_atom.serial_number, atom.alternate_location_indicator)))
TestForceField.py::TestForceField::test_ImpropersOrdering_smirnoff
  /home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/simtk/openmm/app/internal/pdbstructure.py:537: UserWarning: WARNING: duplicate atom (HETATM    3          A   1      -1.098   0.149   0.015  1.00  0.00           H  , HETATM    2          A   1       0.513  -1.098  -0.004  1.00  0.00           O  )
    warnings.warn("WARNING: duplicate atom (%s, %s)" % (atom, old_atom._pdb_string(old_atom.serial_number, atom.alternate_location_indicator)))
TestForceField.py::TestForceField::test_ImpropersOrdering_smirnoff
  /home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/simtk/openmm/app/internal/pdbstructure.py:537: UserWarning: WARNING: duplicate atom (HETATM    4          A   1       0.590   0.938  -0.011  1.00  0.00           H  , HETATM    3          A   1      -1.098   0.149   0.015  1.00  0.00           H  )
    warnings.warn("WARNING: duplicate atom (%s, %s)" % (atom, old_atom._pdb_string(old_atom.serial_number, atom.alternate_location_indicator)))
-- Docs: https://docs.pytest.org/en/latest/warnings.html
====== 1 failed, 276 passed, 14 skipped, 11 warnings in 560.87s (0:09:20) ======
```


* CTest: 6 tests (CPU-related) fail:

```
Errors while running CTest
Test project /home/travis/build/jaimergp/openmm
    Start 46: TestCpuCustomGBForce
1/6 Test #46: TestCpuCustomGBForce .............***Failed    0.06 sec
exception: Assertion failure at TestCustomGBForce.h:153.  Expected -71375.1, found -71348
    Start 49: TestCpuDispersionPME
2/6 Test #49: TestCpuDispersionPME .............***Failed    0.02 sec
exception: Assertion failure at TestDispersionPME.h:1214.  Expected 12.9545, found 12.9517
    Start 50: TestCpuEwald
3/6 Test #50: TestCpuEwald .....................***Failed    0.05 sec
exception: Assertion failure at TestEwald.h:95.  Expected -430767, found -431642
    Start 51: TestCpuGBSAOBCForce
4/6 Test #51: TestCpuGBSAOBCForce ..............***Failed    0.02 sec
exception: Assertion failure at TestGBSAOBCForce.h:237.  Expected 0, found 1.43238
    Start 57: TestCpuNonbondedForce
5/6 Test #57: TestCpuNonbondedForce ............***Failed    0.27 sec
exception: Assertion failure at TestNonbondedForce.h:651.   Expected [-33.1043, -70.073, 13.5967], found [-33.2215, -69.8675, 13.4623]
    Start 79: TestCpuPme
6/6 Test #79: TestCpuPme .......................***Failed    0.69 sec
exception: Assertion failure at TestCpuPme.cpp:635.  Expected -826.349, found -870.677
```

I don't know if this is useful, but trying with `clang` gives one error concerning vector intrinsics as you said, @peastman.  Check [attempt here](https://travis-ci.com/github/conda-forge/openmm-feedstock/jobs/343077061#L3265):

```
In file included from /home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/platforms/cpu/src/CpuCustomGBForce.cpp:30:
In file included from /home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/platforms/cpu/./include/CpuCustomGBForce.h:33:
In file included from /home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/openmmapi/include/openmm/internal/vectorize.h:38:
/home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/openmmapi/include/openmm/internal/vectorize_ppc.h:286:20: error: use of overloaded operator '+' is ambiguous (with operand types 'fvec4' and '__vector float' (vector of 4 'float' values))
    fvec4 temp = r + vec_sld(r.val, r.val, 8);
                 ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~
/home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/openmmapi/include/openmm/internal/vectorize_ppc.h:100:11: note: candidate function
    fvec4 operator+(fvec4 other) const {
          ^
/home/conda/feedstock_root/build_artifacts/openmm_1591177558831/work/openmmapi/include/openmm/internal/vectorize_ppc.h:286:20: note: built-in candidate operator+(__vector float, __vector float)
    fvec4 temp = r + vec_sld(r.val, r.val, 8);
                   ^
```

--- a/openmmapi/include/openmm/internal/vectorize_ppc.h
+++ b/openmmapi/include/openmm/internal/vectorize_ppc.h
@@ -9,7 +9,7 @@
  * Biological Structures at Stanford, funded under the NIH Roadmap for        *
  * Medical Research, grant U54 GM072970. See https://simtk.org.               *
  *                                                                            *
- * Portions copyright (c) 2013-2018 Stanford University and the Authors.      *
+ * Portions copyright (c) 2013-2020 Stanford University and the Authors.      *
  * Authors: Peter Eastman, Heng Ma                                            *
  * Contributors:                                                              *
  *                                                                            *
@@ -66,7 +66,7 @@ public:
     }
     fvec4(__m128 v) : val(v) {}
     fvec4(const float* v) {
-        val = *((__m128*) v);
+        val = (__m128) {v[0], v[1], v[2], v[3]};
     }
 
     /**
@@ -85,7 +85,10 @@ public:
         return val[i];
     }
     void store(float* v) const {
-        *((__m128*) v) = val;
+        v[0] = val[0];
+        v[1] = val[1];
+        v[2] = val[2];
+        v[3] = val[3];
     }
 
     /**
@@ -162,7 +165,7 @@ public:
     }
     ivec4(__m128i v) : val(v) {}
     ivec4(const int* v) {
-        val = *((__m128i*) v);
+        val = (__m128i) {v[0], v[1], v[2], v[3]};
     }
     operator __m128i() const {
         return val;
@@ -171,7 +174,10 @@ public:
         return val[i];
     }
     void store(int* v) const {
-        *((__m128i*) v) = val;
+        v[0] = val[0];
+        v[1] = val[1];
+        v[2] = val[2];
+        v[3] = val[3];
     }
     ivec4 operator+(ivec4 other) const {
         return vec_add(val, other.val);

